### PR TITLE
PowerShell Module for Oracle.Database

### PR DESCRIPTION
**Pull request by v-jamcheung**
## Description

PowerShell Module for Oracle.Database

Merged PR for Oracle.Database Stable Spec: https://github.com/Azure/azure-rest-api-specs/pull/29397, https://github.com/Azure/azure-rest-api-specs/pull/29618

## Mandatory Checklist

- Please choose the target release of Azure PowerShell. (⚠️**Target release** is a different concept from **API readiness**. Please click below links for details.)
  - [x] [General release](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Public preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Private preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Engineering build](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] No need for a release

- [ ] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For SDK-based development mode, update `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`.
        * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. 
    * For autorest-based development mode, include the changelog in the PR description.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Comment by dolauli** [comment]
The root cause of the failures are in typespec-autorest, which is used to generate swaggers 2 from typespec. And I have created an issue for tsp team in https://github.com/Azure/typespec-azure-pr/issues/4235. To fix those failures, you will need to remove those readOnly properties from the required list in the swagger.

**Comment by v-jamcheung** [comment]
> The root cause of the failures are in typespec-autorest, which is used to generate swaggers 2 from typespec. And I have created an issue for tsp team in [Azure/typespec-azure-pr#4235](https://github.com/Azure/typespec-azure-pr/issues/4235). To fix those failures, you will need to remove those readOnly properties from the required list in the swagger.

PR to fix this has been merged: https://github.com/Azure/azure-rest-api-specs/pull/29618
I have regenerated the PowerShell module using the corresponding commitId: https://github.com/Azure/azure-rest-api-specs/commit/eabbc8ab2cc98dc0753ae6eebffc5b4ba5def1ec

**Comment by vsramki** [comment]
> The root cause of the failures are in typespec-autorest, which is used to generate swaggers 2 from typespec. And I have created an issue for tsp team in https://github.com/Azure/typespec-azure-pr/issues/4235. To fix those failures, you will need to remove those readOnly properties from the required list in the swagger.

@dolauli Oracle raised a PR as per your suggestions but ARM review have come back with other suggestions here https://github.com/Azure/typespec-azure-pr/issues/4235#issuecomment-2198222181 . There also seem to be a suggestion to the Powershell SDK team for their checks. Could the powershell team get in touch with the ARM review board (@mikekistler) within Microsoft and then let us (Oracle) know if there needs to be any follow up actions (on both the spec and the powershell PR)? Thank you.

cc: @v-jamcheung  @v-maksymm @v-eelhomsi

**Comment by v-jamcheung** [comment]
> The root cause of the failures are in typespec-autorest, which is used to generate swaggers 2 from typespec. And I have created an issue for tsp team in [Azure/typespec-azure-pr#4235](https://github.com/Azure/typespec-azure-pr/issues/4235). To fix those failures, you will need to remove those readOnly properties from the required list in the swagger.

Issue has been addressed by https://github.com/Azure/azure-rest-api-specs/pull/29618
PR builds are now succeeding.

File: src/OracleDatabase/OracleDatabase.Autorest/examples/Add-AzOracleDatabaseCloudVMClusterVM.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+$dbServerList = Get-AzOracleDatabaseDbServer -Cloudexadatainfrastructurename "OFake_PowerShellTestExaInfra" -ResourceGroupName $resourceGroup
+$dbServerOcid1 = $dbServerList[0].Ocid
+$dbServersToAdd = @($dbServerOcid1)
+
+Add-AzOracleDatabaseCloudVMClusterVM -Cloudvmclustername "OFake_PowerShellTestVmCluster" -ResourceGroupName $resourceGroup -DbServer $dbServersToAdd
+```
+
+Adds a VM to a Cloud VM Cluster resource.
+For more information, execute `Get-Help Add-AzOracleDatabaseCloudVMClusterVM`
```
Comment (YanaXu) [inline src/OracleDatabase/OracleDatabase.Autorest/examples/Add-AzOracleDatabaseCloudVMClusterVM.md:13]:
```suggestion
For more information, execute `Get-Help Add-AzOracleDatabaseCloudVMClusterVM`.
```

File: src/OracleDatabase/OracleDatabase.Autorest/examples/Add-AzOracleDatabaseCloudVMClusterVM.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+$resourceGroup = "PowerShellTestRg"
+
+$dbServerList = Get-AzOracleDatabaseDbServer -Cloudexadatainfrastructurename "OFake_PowerShellTestExaInfra" -ResourceGroupName $resourceGroup
+$dbServerOcid1 = $dbServerList[0].Ocid
+$dbServersToAdd = @($dbServerOcid1)
+
+Add-AzOracleDatabaseCloudVMClusterVM -Cloudvmclustername "OFake_PowerShellTestVmCluster" -ResourceGroupName $resourceGroup -DbServer $dbServersToAdd
+```
+
+Adds a VM to a Cloud VM Cluster resource.
```
Comment (YanaXu) [inline src/OracleDatabase/OracleDatabase.Autorest/examples/Add-AzOracleDatabaseCloudVMClusterVM.md:12]:
```suggestion
Add a VM to a Cloud VM Cluster resource.
```

File: src/OracleDatabase/OracleDatabase.Autorest/examples/Get-AzOracleDatabaseAutonomousDatabase.md [hunk 3]
```diff
@@ -0,0 +1,352 @@
+### Example 1: Gets a list of the Autonomous Database resources
+```powershell
+Get-AzOracleDatabaseAutonomousDatabase
+```
+
+```output
+Location           Name                       SystemDataCreatedAt SystemDataCreatedBy                      SystemDataCreatedByType SystemDataLastModifiedAt SystemDataLastModifiedBy             SystemDataLastModifiedByType ResourceGroupName
+--------           ----                       ------------------- -------------------                      ----------------------- ------------------------ ------------------------             ---------------------------- -----------------
+eastus             ADBScli1                   05/07/2024 13:07:02 ramakrishnan.vilathur.sriniva@oracle.com User                    05/07/2024 13:15:49      857ad006-4380-4712-ba4c-22f7c64d84e7 Application                  SDKTestRG
```
Comment (YanaXu) [inline src/OracleDatabase/OracleDatabase.Autorest/examples/Get-AzOracleDatabaseAutonomousDatabase.md:9]: You can manually update the output and make it shorter.
Don't forget to mask your mails. You can use a fake mail instead.

File: src/OracleDatabase/OracleDatabase.Autorest/test/AzExaInfra.Recording.json [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
...ructures\",\"location\":\"eastus\",\"zones\":[\"2\"],\"systemData\":{\"createdBy\":\"jamie.c.cheung@oracle.com\",\"createdByType\":\"User\",\"createdAt\":\"2024-07-04T13:14:45.3380609Z\",\"lastModifiedBy\":\"jamie.c.cheung@oracle.com\",\"lastModifiedByType\":\"User\",\"lastModifiedAt\":\"2024-07-04T13:14:45.3380609Z\"},\"properties\":{\"ocid\":null,\"computeCount\":3,\"storageCount\":3,\"totalStorageSizeInGbs\":null,\"availableStorageSizeInGbs\":null,\"timeCreated\":null,\"lifecycleDetails\":null,\"maintenanceWindow\":null,\"estimatedPatchingTime\":null,\"customerContacts\":null,\"provisioningState\":\"Accepted\",\"lifecycleState\":null,\"shape\":\"Exadata.X9M\",\"ociUrl\":null,\"displayName\":\"OFake_PowerShellTestExaInfra\",\"cpuCount\":null,\"maxCpuCount\":null,\"memorySizeInGbs\":null,\"maxMemoryInGbs\":null,\"dbNodeStorageSizeInGbs\":null,\"maxDbNodeStorageSizeInGbs\":null,\"dataStorageSizeInTbs\":null,\"maxDataStorageInTbs\":null,\"dbServerVersion\":null,\"activatedStorageCount\":null,\"additionalStorageCount\":null,\"lastMaintenanceRunId\":null,\"nextMaintenanceRunId\":null,\"monthlyDbServerVersion\":null,\"monthlyStorageServerVersion\":null,\"storageServerVersion\":null}}",
```
Comment (YanaXu) [inline src/OracleDatabase/OracleDatabase.Autorest/test/AzExaInfra.Recording.json:40]: Mask your mail in recording files.

**Comment by YanaXu** [comment]
Please run "autorest", "build-module.ps1" again. Some files are not generated, e.g., "New-AzOracleDatabaseCustomerContactObject.md".

File: src/OracleDatabase/OracleDatabase.Autorest/examples/Add-AzOracleDatabaseCloudVMClusterVM.md (hunk 1 above)
Comment (v-jamcheung) [reply src/OracleDatabase/OracleDatabase.Autorest/examples/Add-AzOracleDatabaseCloudVMClusterVM.md]: Updated

File: src/OracleDatabase/OracleDatabase.Autorest/examples/Add-AzOracleDatabaseCloudVMClusterVM.md (hunk 2 above)
Comment (v-jamcheung) [reply src/OracleDatabase/OracleDatabase.Autorest/examples/Add-AzOracleDatabaseCloudVMClusterVM.md]: Updated

File: src/OracleDatabase/OracleDatabase.Autorest/examples/Get-AzOracleDatabaseAutonomousDatabase.md (hunk 3 above)
Comment (v-jamcheung) [reply src/OracleDatabase/OracleDatabase.Autorest/examples/Get-AzOracleDatabaseAutonomousDatabase.md]: Updated

File: src/OracleDatabase/OracleDatabase.Autorest/test/AzExaInfra.Recording.json (hunk 4 above)
Comment (v-jamcheung) [reply src/OracleDatabase/OracleDatabase.Autorest/test/AzExaInfra.Recording.json]: Updated

**Comment by v-jamcheung** [comment]
> Please run "autorest", "build-module.ps1" again. Some files are not generated, e.g., "New-AzOracleDatabaseCustomerContactObject.md".

Updated

File: src/OracleDatabase/OracleDatabase.Autorest/test/Add-AzOracleDatabaseCloudVMClusterVM.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    It 'AddViaJsonString' -skip {
+        { throw [System.NotImplementedException] } | Should -Not -Throw
+    }
+
+    It 'AddViaJsonFilePath' -skip {
+        { throw [System.NotImplementedException] } | Should -Not -Throw
+    }
+
+    It 'Add' -skip {
```
Comment (YanaXu) [inline src/OracleDatabase/OracleDatabase.Autorest/test/Add-AzOracleDatabaseCloudVMClusterVM.Tests.ps1:37]: After you update the README.md file and rerun "autorest" & "build-module.ps1", please remove the test cases for removed parameter sets.

File: src/OracleDatabase/OracleDatabase.Autorest/docs/New-AzOracleDatabaseAutonomousDatabase.md [hunk 6]
```diff
@@ -0,0 +1,860 @@
...
+Aliases:
+
+Required: True
+Position: Named
+Default value: None
+Accept pipeline input: False
+Accept wildcard characters: False
+```
+
+### -ScheduledOperationScheduledStartTime
```
[the remark this answers is not in the document]
Comment (v-jamcheung) [reply src/OracleDatabase/OracleDatabase.Autorest/docs/New-AzOracleDatabaseAutonomousDatabase.md]: Updated

File: src/OracleDatabase/OracleDatabase.Autorest/docs/New-AzOracleDatabaseAutonomousDatabase.md [hunk 7]
```diff
@@ -0,0 +1,860 @@
...
+Aliases:
+
+Required: False
+Position: Named
+Default value: None
+Accept pipeline input: False
+Accept wildcard characters: False
+```
+
+### -ScheduledOperationScheduledStopTime
```
[the remark this answers is not in the document]
Comment (v-jamcheung) [reply src/OracleDatabase/OracleDatabase.Autorest/docs/New-AzOracleDatabaseAutonomousDatabase.md]: Updated

File: src/OracleDatabase/OracleDatabase.Autorest/README.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+module-version: 0.1.0
+title: OracleDatabase
+subject-prefix: $(service-name)
+ 
+inlining-threshold: 100
+resourcegroup-append: true
+nested-object-to-string: true
+identity-correction-for-post: true
+
+directive:
```
[the remark this answers is not in the document]
Comment (v-jamcheung) [reply src/OracleDatabase/OracleDatabase.Autorest/README.md]: Updated

File: src/OracleDatabase/OracleDatabase.Autorest/test/Add-AzOracleDatabaseCloudVMClusterVM.Tests.ps1 (hunk 5 above)
Comment (v-jamcheung) [reply src/OracleDatabase/OracleDatabase.Autorest/test/Add-AzOracleDatabaseCloudVMClusterVM.Tests.ps1]: Updated

File: src/Oracle/Oracle.Autorest/test/Remove-AzOracleOracleSubscription.Tests.ps1 [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  $TestRecordingFile = Join-Path $PSScriptRoot 'Remove-AzOracleOracleSubscription.Recording.json'
+  $currentPath = $PSScriptRoot
+  while(-not $mockingPath) {
+      $mockingPath = Get-ChildItem -Path $currentPath -Recurse -Include 'HttpPipelineMocking.ps1' -File
+      $currentPath = Split-Path -Path $currentPath -Parent
+  }
+  . ($mockingPath | Select-Object -First 1).FullName
+}
+
+Describe 'Remove-AzOracleOracleSubscription' {
```
Comment (YanaXu) [inline src/Oracle/Oracle.Autorest/test/Remove-AzOracleOracleSubscription.Tests.ps1:17]: Please remove this file if the cmdlet is removed.

File: src/Oracle/Oracle.Autorest/docs/Add-AzOracleCloudVMClusterVM.md [hunk 10]
```diff
@@ -0,0 +1,261 @@
+---
+external help file:
+Module Name: Az.Oracle
+online version: https://learn.microsoft.com/powershell/module/az.oracle/add-azoraclecloudvmclustervm
+schema: 2.0.0
+---
+
+# Add-AzOracleCloudVMClusterVM
```
[the remark this answers is not in the document]
Comment (v-jamcheung) [reply src/Oracle/Oracle.Autorest/docs/Add-AzOracleCloudVMClusterVM.md:8]: This command is adding a VM to a Cloud VM Cluster resource - leaving as is.

File: src/Oracle/Oracle.Autorest/README.md [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    - model-name: CustomerContact
+  - model-cmdlet:
+    - model-name: NsgCidr
+
+  # Remove the set-* cmdlet
+  - where:
+      verb: Set
+    remove: true
+
+  # Hide APIs
```
[the remark this answers is not in the document]
Comment (v-jamcheung) [reply src/Oracle/Oracle.Autorest/README.md]: Updated

File: src/Oracle/Oracle.Autorest/test/Add-AzOracleCloudVMClusterVM.Tests.ps1 [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+            $dbServerList = Get-AzOracleDbServer -Cloudexadatainfrastructurename $env.exaInfraName -ResourceGroupName $env.resourceGroup
+            # VM Cluster already contains the first 2 out of 3 VMs from the Exadata Infra
+            $dbServerOcid1 = $dbServerList[2].Ocid
+            $dbServersToAdd = @($dbServerOcid1)
+            
+            Add-AzOracleCloudVMClusterVM -NoWait -Cloudvmclustername $env.vmClusterName -ResourceGroupName $env.resourceGroup -DbServer $dbServersToAdd
+        } | Should -Not -Throw
+    }
+
+    It 'AddViaJsonString' -skip {
```
[the remark this answers is not in the document]
Comment (v-jamcheung) [reply src/Oracle/Oracle.Autorest/test/Add-AzOracleCloudVMClusterVM.Tests.ps1:29]: Updated

File: src/Oracle/Oracle.Autorest/test/Get-AzOracleAutonomousNationalCharacterSet.Tests.ps1 [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  $TestRecordingFile = Join-Path $PSScriptRoot 'Get-AzOracleAutonomousDatabaseNationalCharacterSet.Recording.json'
+  $currentPath = $PSScriptRoot
+  while(-not $mockingPath) {
+      $mockingPath = Get-ChildItem -Path $currentPath -Recurse -Include 'HttpPipelineMocking.ps1' -File
+      $currentPath = Split-Path -Path $currentPath -Parent
+  }
+  . ($mockingPath | Select-Object -First 1).FullName
+}
+
+Describe 'Get-AzOracleAutonomousDatabaseNationalCharacterSet' {
```
[the remark this answers is not in the document]
Comment (v-jamcheung) [reply src/Oracle/Oracle.Autorest/test/Get-AzOracleAutonomousNationalCharacterSet.Tests.ps1]: Updated

File: src/Oracle/Oracle.Autorest/test/Remove-AzOracleOracleSubscription.Tests.ps1 (hunk 9 above)
Comment (v-jamcheung) [reply src/Oracle/Oracle.Autorest/test/Remove-AzOracleOracleSubscription.Tests.ps1]: Updated

**Review by YanaXu — APPROVED** [review]

**Pull request merged**